First PR

paikwiki opened the pull request.

paikwiki committed:
feat: Create main storyboard(draft)

paikwiki committed:
Create an outlet(answerLabel) and an action(touchKeypad)

paikwiki committed:
Edit a storyboard

- Create a table view and a stack view: submit list and keypad

paikwiki committed:
Create an answer when a view did load

paikwiki committed:
Create classes: Answer, Inning

paikwiki committed:
Edit ViewController(WIP)

paikwiki committed:
Done ios-number-baseball Ver.1.0

paikwiki committed:
Edit main storyboard for iPhoneSE(2G)

paikwiki committed:
Do refactor: change a method name

- throwABall() -> pitchABall()
- throw 키워드 때문에 오류를 던진다는 의미로 오해받을 수 있다고 판단해서 변경함

paikwiki committed:
Move code for update labels from Game to ViewController

paikwiki committed:
Do refactor: Change a variable name form inningsData to labels

- 레이블을 모아둔 배열이 이름이 inningsData인 게 부적절하다고 판단함
- labels로 변경

paikwiki committed:
Change variable to constant if it is possible

paikwiki committed:
Remove a param of Game: labels

- 레이블을 업데이트하는 코드는 모두 ViewController에 있음
- Game 클래스가 labels에 대해 알 필요가 없으므로 삭제

paikwiki committed:
Refactor: Remove useless getters

paikwiki committed:
Edit configuration: StatusBarStyle, LaunchScreen, DeviceOrientation

paikwiki committed:
Add a comment "UI UPDATE" on touchKeypad(_:)

rinndash commented on NumberBaseball/LabelsForPitchesAndInningResult.swift:
```diff
@@ -0,0 +1,14 @@
+
+import UIKit
+
+struct LabelsForPitchesAndInningResult {
```
이닝의 숫자와 결과를 묶기 위해 구조체를 쓴 것 같은데 그 용도라면 `InningView`라는 이름으로 `UIView`의 서브클래스를 만들고 그 안에 `UILabel` 두 개를 서브뷰로 추가하는 편이 좋을 듯

paikwiki replied:
수정했습니다.

- 서브뷰(InningView) 적용: [eb8a5c1](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/eb8a5c14ae63fc9172d6733b2533cab84d9f1f2b)

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -9,12 +9,98 @@
-
+    
+    @IBOutlet var answerLabel: UILabel!
+    @IBOutlet var pitchesLabel01: UILabel!
```
위에 남긴 대로 모든 이닝의 숫자와 결과를 UILabel로 뷰컨트롤러에 추가하는 것보다는 `InningView`로 한 번 묶어서 정리하면 좀 더 간단할 듯

paikwiki replied:
수정했습니다.

- 서브뷰(InningView) 적용: [eb8a5c1](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/eb8a5c14ae63fc9172d6733b2533cab84d9f1f2b)

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -9,12 +9,98 @@
+    private var labels = [Int: LabelsForPitchesAndInningResult]()
+    private let answer = Answer()
+    private let inning = Inning()
+    lazy private var game = Game(answer: answer, inning: inning)
```
`lazy`를 사용한 이유가 있나요?

paikwiki replied:
`answer`와 `inning`을 `Game`의 파라미터로 전달할 때 아래처럼 오류가 납니다.

> Cannot use instance member 'answer' within property initializer; property initializers run before 'self' is available
> Cannot use instance member 'inning' within property initializer; property initializers run before 'self' is available

그래서 `self`를 붙여줘도 오류가 나더군요.

책에서 본 기억에 `lazy`를 붙여주면 필요할 때 변수를 메모리에 올린다고 했기 때문에, `lazy`를 이용해 `answer`와 `inning`이 이미 할당된 시점에 `game`을 생성하겠다는 의도로 `lazy`를 썼습니다.

`self`를 쓰려면 `ViewController`가 생성되는 시점에 `answer`와 `inning`을 파라미터로 전달해줘야 할 것 같은데, 그 방법을 몰라서 `lazy` 키워드로 앱이 문제없이 동작하는 것을 확인한 후, `lazy`를 사용하기로 결정했습니다.

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -9,12 +9,98 @@
-
+    
+    @IBAction func touchKeypad(_ sender: UIButton) {
+        let pitchNumber = Character(sender.currentTitle!);
```
애플리케이션의 거의 모든 로직이 이 함수 하나에 다 들어가 있음. 자세한 내용은 아래에 적습니다

paikwiki replied:
다른 세부적인 코드에 대한 리뷰 내용을 수정하면서 `didTapReset(_ sender: UIButton)`과 `didTapNumber(_ sender: UIButton)`, 두 개의 함수로 로직을 분리했습니다.

- `didTapReset(_ sender: UIButton)` 생성 후 키패드의 리셋 버튼과 연결: [ab4f59](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/ab4f595b3677cb2f35b65efbaf587555647b141a)
- 함수명 `touchKeypad(_ sender:UIButton)`을 `didTapNumber(_ sender: UIButton)`으로 변경: [28b863](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/28b8636c9beeeaef8f6b7d441367e4bd96cb8f71)
- `if~return` 구문을 `guard` 구문으로 변경: [a89cb3](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/a89cb343f8bd44d2013332eb3d01569cd448c32f)

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -9,12 +9,98 @@
-
+    
+    @IBAction func touchKeypad(_ sender: UIButton) {
+        let pitchNumber = Character(sender.currentTitle!);
```
옵셔널을 사용할 때 ! 는 최대한 안쓰는 게 좋아. `guard let` 문을 쓰면 !표 없이 작성할 수 있어

```swift
guard let title = sender.currentTitle else { return }
let picthNumber = Character(title)
```

옵셔널의 map을 사용해도 됨

```swift
guard let picthNumber = sender.currentTitle.map(Character.init) else { return }
```

그리고 세미콜론은 안쓰도록 해주세요

paikwiki replied:
수정했습니다.

- 강제언래핑 대신 `guard let` 구문 사용: [ca28b6](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/ca28b636d87b884b44adf6d6f90dbfb2bd518820)
- 세미콜론 제거: [1e51a0](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/1e51a04bb4eb824db64b65c39e6c15552ffb7338)

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -9,12 +9,98 @@
+        let pitchNumber = Character(sender.currentTitle!);
+        
+        // TODO: 빈 버튼 어떤 걸로 사용할 지 결정
+        if pitchNumber == " " {
```
이렇게 return을 바로 하는 경우는 guard 문을 사용하면 코드가 간결합니다

위의 코드에다가  추가한다면 guard 문에 ,를 붙여서 여러 개를 한 번에 쓸 수 있음

```swift
guard 
    let picthNumber = sender.currentTitle.map(Character.init),
    picthNumber != " "
    else { return }
```

paikwiki replied:
수정했습니다.

- `if~return` 구문을 `guard` 구문으로 변경: [a89cb3](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/a89cb343f8bd44d2013332eb3d01569cd448c32f)

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -9,12 +9,98 @@
+        }
+        
+        if pitchNumber == "↺" {
+            // UI UPDATE
```
이 함수에서 버튼을 모두 누를 때를 처리하기보다는 `didTapNumber(_ sender: UIButton)`, `didTapReset(_ sender: UIButton)`으로 분리하는 것이 좋을 듯

이렇게 분리하다면 위에서 빈 버튼을 처리하는 코드도 굳이 없앨 수도 있을 듯

paikwiki replied:
수정했습니다.

- `didTapReset(_ sender: UIButton)` 생성 후 키패드의 리셋 버튼과 연결: [ab4f59](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/ab4f595b3677cb2f35b65efbaf587555647b141a)
- 함수명 `touchKeypad(_ sender:UIButton)`을 `didTapNumber(_ sender: UIButton)`으로 변경: [28b863](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/28b8636c9beeeaef8f6b7d441367e4bd96cb8f71)
- [28b863](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/28b8636c9beeeaef8f6b7d441367e4bd96cb8f71) 커밋의 작업을 하면서 메인스토리보드에서 빈 버튼을 처리하는 액션도 제거했습니다.

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -9,12 +9,98 @@
+            return
+        }
+        
+        if game.isOver {
```
`guard` 문 사용하기

paikwiki replied:
수정했습니다.

- `if~return` 구문을 `guard` 구문으로 변경: [a89cb3](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/a89cb343f8bd44d2013332eb3d01569cd448c32f)

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -9,12 +9,98 @@
+            return
+        }
+        
+        if (inning.pitches.contains(pitchNumber)) {
```
`guard` 사용하기

paikwiki replied:
수정했습니다.

- `if~return` 구문을 `guard` 구문으로 변경: [a89cb3](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/a89cb343f8bd44d2013332eb3d01569cd448c32f)

rinndash commented on NumberBaseball/Inning.swift:
```diff
@@ -0,0 +1,48 @@
+        inningCount = 1
+    }
+    
+    func updatePitchesString(pitchNumber: Character) {
```
`pitchABall` 이후에 이 함수를 호출해 `pitchesString`을 업데이트하기 보다는 `pitchesString`을 계산 프로퍼티로 만드는 편이 좋을 듯.

이 정도 코드는 계산하는 데 드는 비용이 크지 않기 때문에 저장 프로퍼티로 상태관리를 해주는 것보다는 계산 프로퍼티로 그때그때 계산해서 쓰는게 코드를 간결하게 유지할 수 있음

아래 코드를 `var pitchesString: String`로 바꿔 봐~

paikwiki replied:
수정했습니다.

- `updatePitchesString(...)` 메서드 제거, pitchesString을 연산 프로퍼티로 변경: [743391](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/7433913f518d787619343de81bf3067b113158af)

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -0,0 +1,59 @@
+        answer.resetAnswer()
+    }
+    
+    func generateInningResultString() {
```
이 코드도 굳이 저장 프로퍼티를 업데이트하도록 구현하지 말고 현재 상태에 따라 inningResultString이 계산되도록 수정해 봅시다

paikwiki replied:
수정했습니다.

- generateInningResultString() 메서드 제거, inningResultString을 연산 프로퍼티로 변경: [ca28d1e](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/ca28d1eacf228cec268945ce1c50f444d5f58ff0)

rinndash commented on NumberBaseball/Answer.swift:
```diff
@@ -0,0 +1,43 @@
+        answer = createAnswer()
+    }
+    
+    func getAnswerString() -> String {
```
이 함수도 계산 프로퍼티로 만들어 봅시다.

`CustomStringConvertible` 이라는 프로토콜이 있는데 한 번 찾아보고 적용해 보세요

paikwiki replied:
수정했습니다. `CustomStringConvertible` 이거 너무 좋네요!

- `Answer` 클래스에 `CustomStringConvertible` 익스텐션 적용: [121243f](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/121243f94d56e501d4d78ae4f71c627de46134b7)

paikwiki committed:
Remove all semicolons

paikwiki committed:
Replace Forced Unwrapping Optional to guard statement

paikwiki committed:
Create didTapReset(_ sender: UIButton)

- touchKeypad(_ sender: UIButton)에서 리셋 버튼의 동작을 분리
- 입력 키패드의 리셋 버튼과 didTapReset(_ sender: UIButton) 연결

paikwiki committed:
Change a method name: touchKeypad() to didTapNumber()

- touchKeypad(_ sender: UIButton) -> didTapNumber(_ sender: UIButton)

paikwiki committed:
Change if~return to guard statement

paikwiki committed:
Change a property to a computed property: pitchesString

- Remove updatePitchesString(pitchNumber: Character)
- Change pitchesString of Inning to a computed property

paikwiki committed:
Change a property to a computed property: inningResultString

- Remove generateInningResultString() of Game
- Change inningResultString of Game to a computed property

paikwiki committed:
Apply an extension CustomStringConvertible on Answer class

- Remove getAnswerString() of Answer

rinndash commented on NumberBaseball/Inning.swift:
```diff
@@ -0,0 +1,44 @@
+    private(set) var inningCount = 1
+    private(set) var pitches = [Character]()
+    var pitchesString: String {
+        get {
```
계산 프로퍼티를 구현할 때 `get { }` 블럭을 쓰지 않아도 구현할 수 있는데 여기서 사용한 이유가 있나요?

paikwiki replied:
`get {}` 블럭을 사용한 이유는 [꼼꼼한 재은씨의 스위프트 문법편을 보며 공부할 때, 정리해둔 노트](https://github.com/paikwiki/swift-textbook-rubypaper/blob/master/notes/ch08-struct-and-class.md#822-연산-프로퍼티)에 계산 프로퍼티는 아래처럼 쓴다고 정리해둬서 반드시 `get {}` 블럭이 필요한 줄 알았습니다.

```swift
class/struct/enum 객체명 {
  ...
  var 프로퍼티명 : 타입 {
    get {
      필요한 연산 과정
      return 반환값
    }
    set(매개변수명) {
      필요한 연산구문
    }
  }
}
```

수정했습니다.

- 계산 프로퍼티의 `get { }` 블럭 제거 [fe7cc0c](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/fe7cc0c1e17b89442adbb1c7ab5ca3427e8b493a)

rinndash replied:
```swift
class Foo {
  var x: Int = 0
}

class Bar {
  private let foo: Foo = Foo()

  var x: Int {
    get { foo.x }
    set { foo.x = newValue }
  }
}

```

`get { }`, `set { }`은 요렇게 자기 프로퍼티가 아닌데 자기 프로퍼티인 것처럼 숨기고 싶을 때 보통 사용합니다. 하지만 저는 일하면서 사용한 적이 그렇게 많지 않습니다. 왜냐하면 기본적으로 `var`보다는 `let`을 최대한 써야 하기 때문에 `set { }`을 쓰는 경우를 최소화하는 것이 좋기 때문이지요.

`get { }`블럭만 쓰는 경우에는 굳이 추가할 필요가 없기 때문에 생략해도 되고요.

rinndash commented on NumberBaseball/Inning.swift:
```diff
@@ -0,0 +1,44 @@
+                guard let pitchesStringIndex = pitchesCharaterArray.firstIndex(of: "_") else { break }
+                pitchesCharaterArray[pitchesStringIndex] = String.Element(String(pitches[index]))
+            }
+            return (String(pitchesCharaterArray))
```
스위프트 컨벤션은 보통 이렇게 괄호로 감싸지 않습니다

paikwiki replied:
수정했습니다.

- 불필요한 괄호 제거 [a0a9a36](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/a0a9a3665823491b9770fcf6511f579a9f5a9126)

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -0,0 +1,59 @@
+    private(set) var totalInning = 9
+    private(set) var isOver = false
+    var inningResultString: String {
+        get {
```
역시 `get { }` 블럭을 사용할 필요가 없습니다

paikwiki replied:
수정했습니다.

- 계산 프로퍼티의 `get { }` 블럭 제거 [fe7cc0c](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/fe7cc0c1e17b89442adbb1c7ab5ca3427e8b493a)

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -0,0 +1,59 @@
+    
+    func isThreeStrikes() -> Bool {
+        if inning.pitches == answer.answer {
+            return (true)
```
스위프트에서는 일반적으로 `()`를 사용하지 않습니다. 제거해 주세요.

paikwiki replied:
수정했습니다.

- 불필요한 괄호 제거 [a0a9a36](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/a0a9a3665823491b9770fcf6511f579a9f5a9126)

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -9,12 +9,87 @@
-
+    
+    @IBAction func didTapNumber(_ sender: UIButton) {
+        guard let title = sender.currentTitle else { return }
```
나라면 이렇게 붙여썼을 듯

```swift
guard
  let pitchNumber = sender.currentTitle.map(Character.init),
  inning.pitches.contains(pitchNumber) == false,
  game.isOver == false 
  else { return }
```

paikwiki replied:
수정했습니다. `!( )` 보다는 `== false`이 명시적이네요. 변경했습니다.

- `guard` 구문 리팩토링 [927d759](https://github.com/paikwiki/ios-number-baseball/pull/2/commits/927d7598cc93263ded341ab42fdfe33ceb13dd66)

paikwiki committed:
Create InningView

- Remove struct LabelsForPitchesAndInningResult
- Apply a subview for each inning

paikwiki committed:
Remove get statement in a computed property

- Computed properties which have no set statement do not need to get statement

paikwiki committed:
Remove useless parentheses

paikwiki committed:
Refactor guard statement on ViewController

rinndash reviewed (approved):
👍 수고하셨습니다~
